File: linked/19.remove-nth-node-from-end-of-list.go
/*
 * @lc app=leetcode id=19 lang=golang
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	fast := head
	prev := dummy

	for i := 0; i < n; i++ {
		fast = fast.Next
	}

	for fast != nil {
		prev = prev.Next
		fast = fast.Next
	}
	prev.Next = prev.Next.Next
	return dummy.Next

}
// @lc code=end

File: linked/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:  # type:ignore
        dummy = ListNode() # type:ignore
        dummy.next = head
        fast, prev = head, dummy
        for _ in range(n):
            fast = fast.next
        while fast:
            fast = fast.next
            prev = prev.next
        prev.next = prev.next.next
        return dummy.next


# @lc code=end
